src/parser.ts

```diff
@@ -4445,13 +4457,99 @@ export function parseImportExpression(
   if (parser.getToken() === Token.Ellipsis) report(parser, Errors.InvalidSpreadInImport);
 
   const source = parseExpression(parser, context, 1, inGroup, parser.tokenPos, parser.linePos, parser.colPos);
+  const node: ESTree.ImportExpression = {
+    type: 'ImportExpression',
+    source
+  };
+
+  if (context & Context.OptionsNext) {
+    let options: ESTree.Expression | null = null;
+
+    if (parser.getToken() === Token.Comma) {
+      consume(parser, context, Token.Comma);
+
+      if (parser.getToken() !== Token.RightParen) {
+        const expContext = (context | Context.DisallowIn) ^ Context.DisallowIn;
+        options = parseExpression(parser, expContext, 1, inGroup, parser.tokenPos, parser.linePos, parser.colPos);
+      }
+    }
+
+    node.options = options;
+    consumeOpt(parser, context, Token.Comma);
+  }
 
   consume(parser, context, Token.RightParen);
 
-  return finishNode(parser, context, start, line, column, {
-    type: 'ImportExpression',
-    source
-  });
+  return finishNode(parser, context, start, line, column, node);
+}
+
+/**
+ * Parses import attributes
+ *
+ * @param parser Parser object
+ * @param context Context masks
+ * @returns
+ */
+
+export function parseImportAttributes(
+  parser: ParserState,
+  context: Context,
+  specifiers: ESTree.ImportDeclaration['specifiers'] = []
+): ESTree.ImportAttribute[] {
+  consume(parser, context, Token.WithKeyword);
+  consume(parser, context, Token.LeftBrace);
+
+  const attributes: ESTree.ImportAttribute[] = [];
+  const keysContent = new Set<ESTree.Literal['value'] | ESTree.Identifier['name']>();
+  const validJSONImportAttributeBindings =
+    specifiers.length >= 1 &&
+    (specifiers[0].type === 'ImportDefaultSpecifier' ||
+      specifiers[0].type === 'ImportNamespaceSpecifier' ||
+      (specifiers[0].type === 'ImportSpecifier' && specifiers[0].imported.name === 'default'));
+
+  while (parser.getToken() !== Token.RightBrace) {
+    const start = parser.tokenPos;
+    const line = parser.linePos;
+    const column = parser.colPos;
+
+    const key = parseIdentifierOrLiteral(parser, context);
+    consume(parser, context, Token.Colon);
+    const value = parseLiteral(parser, context);
+    const keyContent = key.type === 'Literal' ? key.value : key.name;
+    const isJSONImportAttribute = keyContent === 'type' && value.value === 'json';
+
+    if (isJSONImportAttribute && !validJSONImportAttributeBindings) {
+      report(parser, Errors.InvalidImportBinding);
```

Comment:
The validJSONImportAttributeBindings logic can probably be delayed here. So we don't have the unnecessary cost for non json import.